### DISCUSSION TITLE
Abstract canopy using MeshTree

### DISCUSSION THREAD

**Discussion by gritk** [eradiate/eradiate#309] [Q&A]
I would like to simulate a canopy with different elements, e.g. two types of trees. I would do this with the class MeshTree, but I was not able to do this - TypeError: ("'canopy' must be <class 'eradiate.scenes.biosphere._core.Canopy'>

I have attached my script.

Could you please give me some advice?

[00-test_canopy_abstract_multi_test.txt](https://github.com/eradiate/eradiate/files/10978753/00-test_canopy_abstract_multi_test.txt)

**Comment by leroyvn** [comment]
Hi @gritk, we're indeed missing a tutorial on mesh-based canopies. This is in the pipe. I'm cc'ing @schunkes, who worked with these as well.

The overall idea is that the `Experiment` is expecting a `Canopy` object. The container you need is a [`DiscreteCanopy`](https://eradiate.readthedocs.io/en/stable/rst/reference_api/generated/autosummary/eradiate.scenes.biosphere.DiscreteCanopy.html), which holds a list of [`InstancedCanopyElement`](https://eradiate.readthedocs.io/en/stable/rst/reference_api/generated/autosummary/eradiate.scenes.biosphere.InstancedCanopyElement.html) containers. Each of your `InstancedCanopyElement` wraps, in turn, a `CanopyElement` (*e.g.* `MeshTree` or `LeafCloud`) object.

Please let us know if you need a more concrete example. Admittedly this workflow is convoluted and we plan to drastically simplify this this year.

By the way, if you are planning to perform simulations on large canopies using the CKD mode, you might want to use the latest `main` instead of v0.22.5: we refactored the codebase and the performance of the spectral loop was greatly improved.

**Comment by gritk** [comment]
Hello @leroyvn - thank you for your prompt reply. I will use the latest version and it would be very helpful to have a more concrete example.

**Comment by schunkes** [comment]
Hello @gritk,

I created an example of how to instantate a tree with a spherical crown, using the parameters in the file you provided:

```python
canopy = ertsc.biosphere.DiscreteCanopy(
    instanced_canopy_elements = [
        ertsc.biosphere.InstancedCanopyElement(
            instance_positions = [[0, 2.0, 0]]*ureg.m,
            canopy_element = ertsc.biosphere.AbstractTree(
                leaf_cloud = ertsc.biosphere.LeafCloud.sphere(
                    radius = 2.0 * ureg.m,
                    n_leaves = 10000,
                    leaf_radius = 5.0 * ureg.cm,
                    leaf_reflectance=0.5,
                    leaf_transmittance=0.5,
                ),
                trunk_height=2.0 * ureg.m,
                trunk_radius=0.6 * ureg.m, 
                trunk_reflectance=0.3, 
                leaf_cloud_extra_offset = [0,0,1.5] * ureg.m,
            )
            
        )
    ],
    size = [2, 2, 3.5]*ureg.m
)
```

Rendered with a perspective camera it produces this tree:
![image](https://user-images.githubusercontent.com/59089956/225388919-d92076ad-ef57-47c0-9b69-9bb233bf048e.png)

To place the other trees in the same scene, create multiple `InstancedCanopyElement` objects and append them to the list `instanced_canopy_elements`.

I hope this clarifies the interface. Let me know if you have further questions.

**Reply by gritk** [reply]
Hello @schunkes 

I have followed your advice - but I have not succeeded in creating two objects in the same scene - see code - only the second element of the list is always displayed.

I hope you can help me.

CODE:
```python
tree_1 = ertsc.biosphere.InstancedCanopyElement(           
            instance_positions = [[-2, -2., 0]]*ureg.m,
            canopy_element = ertsc.biosphere.AbstractTree(
                leaf_cloud = ertsc.biosphere.LeafCloud.sphere(
                    radius = 2.0 * ureg.m,
                    n_leaves = 10000,
                    leaf_radius = 5.0 * ureg.cm,
                    leaf_reflectance=0.5,
                    leaf_transmittance=0.5,
                ),
                trunk_height=2.0 * ureg.m,
                trunk_radius=0.6 * ureg.m, 
                trunk_reflectance=0.3, 
                leaf_cloud_extra_offset = [0,0,1.5] * ureg.m,
            )
            
        )
tree_2 = ertsc.biosphere.InstancedCanopyElement(
            instance_positions = [[5.0, 5.0, 0]]*ureg.m,
            canopy_element = ertsc.biosphere.AbstractTree(
                leaf_cloud = ertsc.biosphere.LeafCloud.sphere(
                    radius = 3.0 * ureg.m,
                    n_leaves = 10000,
                    leaf_radius = 5.0 * ureg.cm,
                    leaf_reflectance=0.5,
                    leaf_transmittance=0.5,
                ),
                trunk_height=2.0 * ureg.m,
                trunk_radius=0.6 * ureg.m, 
                trunk_reflectance=0.3, 
                leaf_cloud_extra_offset = [0,0,1.5] * ureg.m,
            )
            
        )

canopy = ertsc.biosphere.DiscreteCanopy(
    instanced_canopy_elements = [tree_2,tree_1],
    size = [20, 20, 20]*ureg.m
)
```

**Reply by schunkes** [reply]
Hello @gritk ,

Sorry for this confusion. If you define multiple trees they need to have unique identifiers. 
The identifiers are set in the canopy element, in this case the `AbstractTree` classes.
I updated your example and this renders both trees correctly:

```python
tree_1 = ertsc.biosphere.InstancedCanopyElement(  
            instance_positions = [[-2, -2., 0]]*ureg.m,
            canopy_element = ertsc.biosphere.AbstractTree(
                # add the ID here
                id="tree_1",
                leaf_cloud = ertsc.biosphere.LeafCloud.sphere(
                    radius = 2.0 * ureg.m,
                    n_leaves = 10000,
                    leaf_radius = 5.0 * ureg.cm,
                    leaf_reflectance=0.5,
                    leaf_transmittance=0.5,
                ),
                trunk_height=2.0 * ureg.m,
                trunk_radius=0.6 * ureg.m, 
                trunk_reflectance=0.3, 
                leaf_cloud_extra_offset = [0,0,1.5] * ureg.m,
            )
            
        )
tree_2 = ertsc.biosphere.InstancedCanopyElement(
            instance_positions = [[5.0, 5.0, 0]]*ureg.m,
            canopy_element = ertsc.biosphere.AbstractTree(
                # add the ID here
                id="tree_2",
                leaf_cloud = ertsc.biosphere.LeafCloud.sphere(
                    radius = 3.0 * ureg.m,
                    n_leaves = 10000,
                    leaf_radius = 5.0 * ureg.cm,
                    leaf_reflectance=0.5,
                    leaf_transmittance=0.5,
                ),
                trunk_height=2.0 * ureg.m,
                trunk_radius=0.6 * ureg.m, 
                trunk_reflectance=0.3, 
                leaf_cloud_extra_offset = [0,0,1.5] * ureg.m,
            )
            
        )

canopy = ertsc.biosphere.DiscreteCanopy(
    instanced_canopy_elements = [tree_2,tree_1],
    size = [20, 20, 20]*ureg.m
)
```

**Comment by gritk** [comment]
Man thanks for your prompt reply - Now it is working as expected!!!
